### ISSUE TITLE
"Queued" jobs not processed after worker connects

### ISSUE THREAD

**Issue by olalonde** [Automattic/kue#926]
<img width="91" alt="screenshot 2016-08-03 14 07 05" src="https://cloud.githubusercontent.com/assets/416585/17381996/a0c0872a-5983-11e6-8d6f-5bcb30da0287.png">

If I click on the "refresh" icon on "Queued" items, they are processed but otherwise they just sit there doing nothing. For some reason, most tasks are processed instantly but some will just sit there in "Queued" doing nothing forever until I click the "refresh" icon. Any idea of what might be causing this?

**Comment by victusfate** [comment]
I had this for a backlog of 1000 jobs, I ended up writing a script to clear them all out and restarted the server.

Not sure what caused it to happen but there were a number of jobs created simultaneously (in error) and each of them is long running.

sample code to start cleaning stuff up in redis (please disregard the sp0n stuff it's from private repo):

```
'use strict';

const config        = require('sp0n-config');
const logglyToken   = config.Loggly.Token;
const subdomain     = config.Loggly.Subdomain;
const logInitialize = require('sp0n-logger').init;
const kue           = require('kue');
const url           = require('url');
const redis         = require('redis');
const util          = require('sp0n-util');

var getRedis = () => {
  // if need to clean up locally process.env.LOCAL_REDIS_URL
  var redisUrl = url.parse(config.Redis.RedisUrl)
    , client = redis.createClient(redisUrl.port, redisUrl.hostname);
  // console.log({ action: 'app', redisUrl: redisUrl });
  if (redisUrl.auth) {
    client.auth(redisUrl.auth.split(":")[1]);
  }
  return client;
}

var isInt = (val) => {
  return typeof val === 'number' && Math.floor(val) === val;
}

var queue = kue.createQueue({
  redis: { 
    createClientFactory: getRedis
  }
});

queue.inactive( function( err, ids ) { // others are active, complete, failed, delayed
  const sAction = 'queue.inactive';
  console.log({ action: sAction, ids: ids });
  for (let id in ids) {
    kue.Job.remove( ids[id], (err) => {
      if (err) {
        console.error({ action: sAction + '.err', id: ids[id], err:err });
      }
    });
  }
});

// another way to go
// queue.Job.rangeByState( 'inactive', 0, 10000, 'asc', function( err, jobs ) {
//   for (let job in jobs) {
//     kue.Job.remove( jobs[job].id, (err) => {
//       if (err) {
//         console.error({ action: sAction + '.err', id: ids[id], err:err });
//       }
//     });
//   }
// });
```

**Comment by olalonde** [comment]
Thanks for the script. I added `queue.watchStuckJobs(5 * 1000)` and problem hasn't re-occurred so far.

**Comment by victusfate** [comment]
We ran into again. It appears to occur whenever a large number of jobs are queued in a short time interval. I'll try to focus on a test case locally and determine where its getting stuck.

@behrad mentioned some changes with the next release, and they may have some effect so I'll test latest if I can reproduce, and see if that resolves the stuck in queued state issue.

**Comment by olalonde** [comment]
Yeah, reproducing the bug would be helpful. Additionally, something super useful would be a CLI vs the web dashboard / the redis REPL. `kue list` `kue retry failed` `kue clear failed` etc. Might work on this when I have time.

**Comment by victusfate** [comment]
I didn't even think redis repl, cool.

**Comment by victusfate** [comment]
Reviewing tips for preventing stuck queues:
https://github.com/Automattic/kue#prevent-from-stuck-active-jobs

This could definitely be the cause as I'm doing media processing and all kinds of interesting errors can arise. I'll go with the domain wrapper or promise setup (all the rest of the code I'm using is promises).

**Comment by victusfate** [comment]
some good stuff mentioned in this thread as well (similar issue)
https://github.com/Automattic/kue/issues/130

I'm trying something in the workers now to gracefully shutdown. I was reliably getting the queue stuck by queueing jobs and killing the worker. It would never have a chance to call done. Also if a worker crashes the same can happen (hence the domain grabs or maybe a try catch)

**Comment by victusfate** [comment]
ahoy olalonde -> this was an earlier batch job I was able to run and kill and get consistently stuck jobs. Now with some modifications it doesn't get stuck, but I'm seeing some active jobs just hanging out in limbo. I commented on this in issue #130 

Ok put together a gist with graceful queue and worker shutdown. I'm still seeing a stock active job, so I think worker pause is not triggering active jobs into an inactive state.

Here's the gist:
https://gist.github.com/victusfate/1e2ce9eb73de32b78d2690d660f0f9c8

**Comment by victusfate** [comment]
Updated the gist to handle setting active jobs to inactive

**Comment by olalonde** [comment]
@victusfate what happens if the process events are not attached though? do jobs get unstuck eventually?

**Comment by victusfate** [comment]
Yeah I didn't handle uncaught exceptions, and there could be other signals I missed but it worked very well while I killed and restarted it testing locally. No stuck queue - in the earlier version I could reliably recreate a stuck queue just be killing the workers and rerunning them.

**Comment by aiavci** [comment]
> I had this for a backlog of 1000+ jobs, I ended up writing a script to clear them all out and restarted the server.
> 
> Not sure what caused it to happen but there were a number of jobs created simultaneously (in error) and each of them is long running.
> 
> sample code to start cleaning stuff up in redis (please disregard the sp0n stuff it's from private repo):
> 
> ```
> 'use strict';
> 
> const config        = require('sp0n-config');
> const logglyToken   = config.Loggly.Token;
> const subdomain     = config.Loggly.Subdomain;
> const logInitialize = require('sp0n-logger').init;
> const kue           = require('kue');
> const url           = require('url');
> const redis         = require('redis');
> const util          = require('sp0n-util');
> 
> var getRedis = () => {
>   // if need to clean up locally process.env.LOCAL_REDIS_URL
>   var redisUrl = url.parse(config.Redis.RedisUrl)
>     , client = redis.createClient(redisUrl.port, redisUrl.hostname);
>   // console.log({ action: 'app', redisUrl: redisUrl });
>   if (redisUrl.auth) {
>     client.auth(redisUrl.auth.split(":")[1]);
>   }
>   return client;
> }
> 
> var isInt = (val) => {
>   return typeof val === 'number' && Math.floor(val) === val;
> }
> 
> var queue = kue.createQueue({
>   redis: { 
>     createClientFactory: getRedis
>   }
> });
> 
> queue.inactive( function( err, ids ) { // others are active, complete, failed, delayed
>   const sAction = 'queue.inactive';
>   console.log({ action: sAction, ids: ids });
>   for (let id in ids) {
>     kue.Job.remove( ids[id], (err) => {
>       if (err) {
>         console.error({ action: sAction + '.err', id: ids[id], err:err });
>       }
>     });
>   }
> });
> 
> // another way to go
> // queue.Job.rangeByState( 'inactive', 0, 10000, 'asc', function( err, jobs ) {
> //   for (let job in jobs) {
> //     kue.Job.remove( jobs[job].id, (err) => {
> //       if (err) {
> //         console.error({ action: sAction + '.err', id: ids[id], err:err });
> //       }
> //     });
> //   }
> // });
> ```

Does this mean jobs that fail are removed and never run?
